lib/charms/alertmanager_k8s/v0/alertmanager_remote_configuration.py

```diff
@@ -0,0 +1,408 @@
+    pass
+
+
+class AlertmanagerRemoteConfigurationRequirerEvents(ObjectEvents):
+    """Event descriptor for events raised by `AlertmanagerRemoteConfigurationRequirer`."""
+
+    remote_configuration_changed = EventSource(AlertmanagerRemoteConfigurationChangedEvent)
+
+
+class RemoteConfigurationRequirer(Object):
+    """API that manages a required `alertmanager_remote_configuration` relation.
+
+    The `RemoteConfigurationRequirer` object can be instantiated as follows in your charm:
+
+    ```
+    from charms.alertmanager_k8s.v0.alertmanager_remote_configuration import (
+        RemoteConfigurationRequirer,
+    )
+
+    def __init__(self, *args):
+        ...
+        self.remote_configuration = RemoteConfigurationRequirer(self)
+        ...
+    ```
+
+    The `RemoteConfigurationRequirer` assumes that, in the `metadata.yaml` of your charm,
+    you declare a required relation as follows:
+
+    ```
+    requires:
+        remote-configuration:  # Relation name
+            interface: alertmanager_remote_configuration  # Relation interface
+            limit: 1
+    ```
+
+    The `RemoteConfigurationRequirer` provides a public `config` method for exposing the data
+    from the relation data bag. Typical usage of these methods in the provider charm would look
+    something like:
+
+    ```
+    def get_config(self, *args):
+        ...
+        configuration, templates = self.remote_configuration.config()
+        ...
+        self.container.push("/alertmanager/config/file.yml", configuration)
+        self.container.push("/alertmanager/templates/file.tmpl", templates)
+        ...
+    ```
+
+    Separation of the main configuration and the templates is dictated by the assumption that
+    the default provider of the `alertmanager_remote_configuration` relation will be
+    `alertmanager-k8s` charm, which requires such separation.
+    """
+
+    on = AlertmanagerRemoteConfigurationRequirerEvents()
+
+    def __init__(
+        self,
+        charm: CharmBase,
+        relation_name: str = DEFAULT_RELATION_NAME,
+    ):
+        """API that manages a required `remote-configuration` relation.
+
+        Args:
+            charm: The charm object that instantiated this class.
+            relation_name: Name of the relation with the `alertmanager_remote_configuration`
+                interface as defined in metadata.yaml. Defaults to `remote-configuration`.
+        """
+        super().__init__(charm, relation_name)
+        self._charm = charm
+        self._relation_name = relation_name
+
+        on_relation = self._charm.on[self._relation_name]
+
+        self.framework.observe(on_relation.relation_changed, self._on_relation_changed)
+
+    def _on_relation_changed(self, _) -> None:
+        """Event handler for remote configuration relation changed event.
+
+        Emits custom `remote_configuration_changed` event every time remote configuration
+        changes.
+        """
+        self.on.remote_configuration_changed.emit()
```

Comment:
It could, but IMO that would require a lot of refactoring in the charm. 
Right now, most (if not all) of the callback functions in the charm run `_common_exit_hook`. That includes `_on_remote_configuration_changed`. If we want to keep it this way, we'd have to push `event` object all the way down to `_update_config` and even then we'd need to refactor the way config is fetched in `alertmanager-k8s`.
Alternative could be running `_update_config` directly on `_on_remote_configuration_changed`, but still refactoring of how the configuration is fetched would be required.
Do you see any added value in changing the way of we fetch the remote configuration?

Reply:
The idea was to reduce the back and forth between the charm and the lib, but if it complicates things then it may not be worth it.